File: ADF/adf.py
# Kuntal Ghosh
# Code for computing adngular distribution functions (ADFs)
# May 2022

import numpy as np

# Constants
dr = 0.02
dang = 1.0
sigma = 10.0
r_fixed = 3.0
pi = np.pi

def angle(v1, v2):
    cos_theta = np.dot(v1, v2)
    d1 = np.linalg.norm(v1)
    d2 = np.linalg.norm(v2)
    cos_theta = cos_theta / (d1 * d2)
    a = np.arccos(cos_theta)
    return np.degrees(a)

# Read the input file
with open('../traj.lammpstrj', 'r') as f:
    lines = f.readlines()

nlines = len(lines)
natoms = int(lines[3])
box_length = np.array([float(line.split()[1]) for line in lines[5:8]])
print(box_length)

rho = natoms / np.prod(box_length)
nframes = nlines // (natoms + 9)

# Initialize arrays
xyz = np.zeros((3, natoms, nframes))
force_xyz = np.zeros((3, natoms, nframes))

# Store coordinates
for i in range(nframes):
    start = i * (natoms + 9) + 9
    for k in range(natoms):
        line = lines[start + k].split()
        xyz[:, k, i] = [float(x) for x in line[1:4]]
        force_xyz[:, k, i] = [float(x) for x in line[4:7]]

angmin, angmax = 0.0, 180.0
dmin, dmax = 0.0, 0.5 * box_length[0]

print(f"angmin, angmax, dmin, dmax: {angmin}, {angmax}, {dmin}, {dmax}")

nbin_ang = int((angmax - angmin) / dang) + 1
a = np.zeros(nbin_ang)

total = 0
for iframe in range(nframes):
    for i in range(natoms):
        for j in range(natoms):
            if i != j:
                v1 = xyz[:, j, iframe] - xyz[:, i, iframe]
                v1 -= box_length * np.round(v1 / box_length)
                d1 = np.dot(v1, v1)

                for k in range(j + 1, natoms):
                    if k != i:
                        v2 = xyz[:, k, iframe] - xyz[:, i, iframe]
                        v2 -= box_length * np.round(v2 / box_length)
                        d2 = np.dot(v2, v2)

                        if d1 < sigma and d2 < sigma:
                            ang = angle(v1, v2)

                            if 0.0 < ang < 180.0:
                                ibin_ang = int((ang - angmin) / dang)

                                if 0 <= ibin_ang < nbin_ang:
                                    a[ibin_ang] += 1.0
                                    total += 1

# Save ADF data
ang_values = np.arange(angmin, angmax + dang, dang)
adf = a / (total * dang)
np.savetxt('adf.dat', np.column_stack((np.radians(ang_values), adf)))

File: ADF/adf.f90
      PROGRAM three_body

      ! Finds the ADF

      ! Kuntal Ghosh
      ! February 2022
      
      IMPLICIT NONE

      INTEGER*8 :: i,j,k,ios,nlines,iframe,nframes,natoms,id
      INTEGER*8 :: ibin_r,ibin_ang,nbin_r,nbin_ang,total
      REAL*8, PARAMETER :: dr=0.02d0, dang=1.0d0,sigma=10.0d0,& 
                           r_fixed=3.0d0,pi=4.0d0*ATAN(1.0d0)
      REAL*8, ALLOCATABLE :: xyz(:,:,:),force_xyz(:,:,:),p(:,:),a(:)
      REAL*8 :: v(3),v1(3),v2(3),dmin,dmax,d,angmin,angmax, &
                ang,rij,arb,box_length(3),norm_fact,rho,d1,d2

      OPEN (1, FILE = '../../cg.lammpstrj', STATUS = 'OLD')
      OPEN (2, FILE = 'adf.dat', STATUS = 'UNKNOWN')

!     Reading the file
      
      nlines = 0
      read_loop: DO
         READ (1,*,IOSTAT=ios)
         IF (ios/=0) EXIT read_loop
         nlines = nlines + 1
      END DO read_loop

      REWIND (1)

      READ (1,*)
      READ (1,*)
      READ (1,*)
      READ (1,*) natoms
      READ (1,*)
      DO i = 1,3
         READ (1,*) arb,box_length(i) !Reads the box length
      END DO
      WRITE (*,*) box_length(1:3)

      REWIND (1)

      rho = natoms/(box_length(1)*box_length(2)*box_length(3))
      nframes = nlines/(natoms + 9) ! Typically, skip the first 9 lines of each frame

      ALLOCATE (xyz(3,natoms,nframes),force_xyz(3,natoms,nframes))

!     Storing all the coordinates

      DO i = 1,nframes
         DO j = 1,9
            READ (1,*)
         END DO
         DO k = 1,natoms
            READ (1,*) id,xyz(1:3,k,i),force_xyz(1:3,k,i)
         END DO
      END DO

      CLOSE (1)


      angmin = 0.0d0
      angmax = 180.0d0
      dmin = 0.0d0
      dmax = 0.5d0*box_length(1)

      WRITE (*,*) "angmin, angmax, dmin, dmax", angmin,angmax,dmin,dmax

      nbin_ang = INT((angmax-angmin)/dang) + 1

      ALLOCATE (p(nbin_r,nbin_ang),a(nbin_ang))

!     Computing a(r) !Angular distribution function

      a(:) = 0.0d0

      DO iframe = 1,nframes
         DO i = 1,natoms
            DO j = 1,natoms
               IF (i/=j) THEN

                  v1(1:3) = xyz(1:3,j,iframe) - xyz(1:3,i,iframe)                  
                  v1(1:3) = v1(1:3) - box_length(1:3)*NINT(v1(1:3)/box_length(1:3))
                  d1 = DOT_PRODUCT(v1,v1)

                  DO k = j+1,natoms
                     IF (k/=i) THEN

                        v2(1:3) = xyz(1:3,k,iframe) - xyz(1:3,i,iframe)                        
                        v2(1:3) = v2(1:3) - box_length(1:3)*NINT(v2(1:3)/box_length(1:3))
                        d2 = DOT_PRODUCT(v2,v2)
                        
                        IF (d1<sigma .AND. d2<sigma) THEN
                           CALL angle (v1,v2,ang)

                           IF (ang>0.0d0 .AND. ang<180.0d0) THEN
                              ibin_ang = INT((ang-angmin)/dang) + 1
                           END IF

                           IF (ibin_ang>0 .AND. ibin_ang<=nbin_ang) THEN
                              a(ibin_ang) = a(ibin_ang) + 1.0d0
                              total = total + 1
                           END IF

                        END IF

                     END IF
                  END DO 
                  
               END IF
            END DO
         END DO
      END DO
         
      ang = angmin
      DO ibin_ang = 1,nbin_ang
         WRITE (2,*) (ang*pi/180.0d0),a(ibin_ang)/(DFLOAT(total)*dang)
         ang = ang + dang
      END DO

      CLOSE (3)


      DEALLOCATE (xyz,force_xyz,p,a)

      END PROGRAM three_body

      


      SUBROUTINE angle (v1,v2,a)

      IMPLICIT NONE

      REAL*8 :: cos_theta,v1(3),v2(3),d1,d2,a
      REAL*8, PARAMETER :: pi = 4.0d0*ATAN(1.0d0)

      cos_theta = DOT_PRODUCT(v1,v2)
      d1 = DSQRT(DOT_PRODUCT(v1,v1))
      d2 = DSQRT(DOT_PRODUCT(v2,v2))
      cos_theta = cos_theta/(d1*d2)
      a = DACOS(cos_theta)
      a = (a*180.0d0)/pi

      END SUBROUTINE angle   
